src/main/java/org/broadinstitute/hellbender/tools/walkers/filters/VariantFiltration.java

```diff
@@ -0,0 +1,389 @@
+                        }
+                    }
+
+                    // if sample is filtered and --setFilteredGtToNocall, set genotype to non-call
+                    if ( !filters.isEmpty() && setFilteredGenotypesToNocall ) {
+                        genotypes.add(new GenotypeBuilder(g).filters(filters).alleles(diploidNoCallAlleles).make());
+                    } else {
+                        genotypes.add(new GenotypeBuilder(g).filters(filters).make());
+                    }
+                } else {
+                    genotypes.add(g);
+                }
+            }
+
+            builder.genotypes(genotypes);
+        }
+
+        // make a new variant context based on filters
+        final Set<String> filters = new LinkedHashSet<>(vc.getFilters());
+
+        // test for clustered SNPs if requested
+        if (areClusteredVariants(featureContext, vc)){
+            filters.add(CLUSTERED_SNP_FILTER_NAME);
+        }
+
+        for ( final JexlVCMatchExp exp : filterExps ) {
+            try {
+                if ( Utils.invertLogic(VariantContextUtils.match(vc, exp), invertFilterExpression) ) {
+                    filters.add(exp.name);
+                }
+            } catch (final Exception e) {
+                // do nothing unless specifically asked to; it just means that the expression isn't defined for this context
+                if ( failMissingValues  ) {
+                    filters.add(exp.name);
+                }
+            }
+        }
+
+        if ( filters.isEmpty() ) {
+            builder.passFilters();
+        } else {
+            builder.filters(filters);
+        }
+
+        writer.add(builder.make());
+    }
+
+    /**
+     * Return true if there is a window of size {@link clusterWindow} that contains as least {@link clusterSize} variants.
+     */
+    private boolean areClusteredVariants(final FeatureContext featureContext, final VariantContext current) {
+        if (clusterWindow < 1){ //as per argument doc, values < 1 imply no clustering
+            return false;
+        }
+        //Need to fetch variants from left and right of the current variant.
+        final List<? extends Feature> values = featureContext.getValues(getDrivingVariantFeatureInput(), clusterWindow, clusterWindow);
+        values.sort(IntervalUtils.LEXICOGRAPHICAL_ORDER_COMPARATOR); //Note: by construction, they will be from the same contig.
+        if (values.size() < clusterSize){  //not enough variants - will never be a cluster no matter what.
+            return false;
+        }
+
+        //Scan the list of variants and check if any stretch of clusterSize variants
+        //that contains the 'current variant' has is more clustered than clusterWindow.
+        final int currentStartPos = current.getStart();
+        int firstIndexIndex = 0;
+
+        //start with the lowest pos
+        int firstPos = values.get(firstIndexIndex).getStart();//there is at least 1 element so it's ok
```

Comment:
doc only says this works for snps. so i'm going to make it work for SNPs only. I can add a ticket to extend to non-SNPs